File: Leetcode2023/April2023/Python/SuccessfulPairsOfSpellsAndPotions.py
class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        n = len(spells)
        m = len(potions)
        potions.sort()
        
        for i in range(n):
            start = 0
            end = m - 1
            ind = -1
            
            while start <= end:
                mid = start + (end - start) // 2
                product = spells[i] * potions[mid]
                
                if product >= success:
                    ind = mid
                    end = mid - 1
                else:
                    start = mid + 1
                    
            spells[i] = m - ind if ind != -1 else 0
        
        return spells
        

File: Leetcode2023/April2023/C++/SuccessfulPairsOfSpellsAndPotions.cpp
class Solution {
public:
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success) {
        int n = spells.size();
        int m = potions.size();
        sort(potions.begin(), potions.end());
        
        for(int i = 0; i < n; i++) {
            int start = 0;
            int end = m - 1;
            int ind = -1;
            
            while(start <= end) {
                int mid = start + (end - start) / 2;
                long long product = (long long)spells[i] * potions[mid];
                
                if(product >= success) {
                    ind = mid;
                    end = mid - 1;
                } else {
                    start = mid + 1;
                }
            }
            
            spells[i] = ind != -1 ? m - ind : 0;
        }
        
        return spells;
    }
};
